servers/rendering/shader_api_dump.cpp

```diff
@@ -0,0 +1,385 @@
+
+				DataType return_type = DataType::TYPE_VOID;
+				bool unary = false;
+				bool valid = ShaderLanguage::get_datatype_operator_result(type, op, other_type, &return_type, &unary);
+				if (valid) {
+					ShaderApiOperator api_operator = {
+						get_operator_name(op),
+						return_type,
+						unary ? DataType::TYPE_VOID : other_type,
+					};
+					operators.push_back(api_operator);
+				}
+
+				// If unary, don't iterate through every "other" type.
+				if (unary) {
+					break;
+				}
+
+				other_type_index++;
+			}
+
+			operator_index++;
+		}
+
+		ShaderApiDatatype datatype = {
+			ShaderLanguage::get_datatype_name(type),
+			ShaderLanguage::get_datatype_indexed_type(type),
+			operators,
+			ShaderLanguage::get_datatype_size(type),
+			ShaderLanguage::get_datatype_component_count(type),
+			ShaderLanguage::get_datatype_indexed_size(type),
+			ShaderLanguage::get_datatype_swizzle_field_count(type),
+			ShaderLanguage::is_scalar_type(type),
+			ShaderLanguage::is_float_type(type),
+			ShaderLanguage::is_sampler_type(type),
+		};
+		datatypes.push_back(datatype);
+	}
+
+	return datatypes;
+}
+
+Vector<GDShaderAPIDump::ShaderApiFunction> GDShaderAPIDump::generate_functions() {
+	HashMap<String, Vector<ShaderApiFunctionOverload>> function_overloads;
+
+	HashMap<String, Vector<int>> out_args = collect_out_arguments();
+	HashMap<String, const ShaderLanguage::BuiltinFuncConstArgs *> const_int_args = collect_const_int_arguments();
+
+	int idx = 0;
+	while (ShaderLanguage::builtin_func_defs[idx].name) {
+		const ShaderLanguage::BuiltinFuncDef &def = ShaderLanguage::builtin_func_defs[idx];
+
+		String function_name = String(def.name);
+
+		// arguments
+		Vector<ShaderApiFunctionArgument> arguments;
+		int argument_idx = 0;
+		const Vector<int> *out_arg_indexes = out_args.getptr(function_name);
+		const ShaderLanguage::BuiltinFuncConstArgs **const_int_arg = const_int_args.getptr(function_name);
+		while (def.args[argument_idx] != ShaderLanguage::DataType::TYPE_VOID) {
+			bool is_out = out_arg_indexes != nullptr ? out_arg_indexes->has(argument_idx) : false;
+
+			const ShaderLanguage::BuiltinFuncConstArgs *const_data = nullptr;
+			if (const_int_arg != nullptr && (*const_int_arg)->arg == argument_idx) {
+				const_data = (*const_int_arg);
+			}
+
+			ShaderApiFunctionArgument argument = {
+				String(def.args_names[argument_idx]),
+				def.args[argument_idx],
+				is_out,
+				const_data
+			};
+			arguments.push_back(argument);
+
+			argument_idx++;
+		}
+
+		// add overload to function
+		ShaderApiFunctionOverload overload = {
+			arguments,
+			def.rettype,
+			def.high_end
+		};
+
+		if (function_overloads.has(function_name)) {
+			function_overloads.get(function_name).push_back(overload);
+		} else {
+			function_overloads.insert(function_name, Vector({ overload }));
+		}
+
+		idx++;
+	}
+
+	Vector<ShaderApiFunction> functions;
+
+	HashSet<String> frag_only_functions = collect_frag_only_functions();
+	for (auto &f : function_overloads) {
```

Comment:
`auto` is [not allowed](https://contributing.godotengine.org/en/latest/engine/guidelines/cpp_usage_guidelines.html#auto-keyword), for all